test/horizontal_fusion_test.cpp

```diff
@@ -0,0 +1,76 @@
+#include <migraphx/literal.hpp>
+#include <migraphx/operators.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/check_shapes.hpp>
+#include <migraphx/generate.hpp>
+#include <migraphx/instruction.hpp>
+#include <basic_ops.hpp>
+#include <test.hpp>
+
+struct horizontal_fusion_target
+{
+    std::string name() const { return "horizontal fusion"; }
+    std::vector<migraphx::pass> get_passes(migraphx::context&) const
+    {
+        return {migraphx::horizontal_fusion{}};
+    }
+    migraphx::context get_context() const { return {}; }
+};
+
+TEST_CASE(test1)
+{
+    migraphx::program p;
+    int size = 192 * 35 * 35;
+    std::vector<float> a(size);
+    for(auto i = 0; i < size; i++)
+        a[i] = 1.0 * i;
+    size = 64 * 192;
+    std::vector<float> b(size);
+    for(auto i = 0; i < size; i++)
+        b[i] = 0.5 * i;
+    size = 64 * 35 * 35;
+    std::vector<float> c(size);
+    for(auto i = 0; i < size; i++)
+        c[i] = 0.1 * i;
+    size = 48 * 192;
+    std::vector<float> d(size);
+    for(auto i = 0; i < size; i++)
+        d[i] = 0.2 * i;
+    size = 48 * 35 * 35;
+    std::vector<float> e(size);
+    for(auto i = 0; i < size; i++)
+        e[i] = 0.3 * i;
+
+    migraphx::shape a_shape{migraphx::shape::float_type, {1, 192, 35, 35}};
+    auto l1 = p.add_literal(migraphx::literal{a_shape, a});
+    auto p1 = p.add_instruction(migraphx::op::pooling{}, l1);
+    migraphx::shape b_shape{migraphx::shape::float_type, {64, 192, 1, 1}};
+    auto l2 = p.add_literal(migraphx::literal{b_shape, b});
+    auto c1 = p.add_instruction(migraphx::op::convolution{}, p1, l2);
+    migraphx::shape c_shape{migraphx::shape::float_type, {1, 64, 35, 35}};
+    auto l3 = p.add_literal(migraphx::literal{c_shape, c});
+    auto a1 = p.add_instruction(migraphx::op::add{}, c1, l3);
+    auto r1 = p.add_instruction(migraphx::op::relu{}, a1);
+    migraphx::shape d_shape{migraphx::shape::float_type, {48, 192, 1, 1}};
+    auto l4 = p.add_literal(migraphx::literal{d_shape, d});
+    auto c2 = p.add_instruction(migraphx::op::convolution{}, p1, l4);
+    migraphx::shape e_shape{migraphx::shape::float_type, {1, 48, 35, 35}};
+    auto l5 = p.add_literal(migraphx::literal{e_shape, e});
+    auto a2 = p.add_instruction(migraphx::op::add{}, c2, l5);
+    auto r2 = p.add_instruction(migraphx::op::relu{}, a2);
+    auto p2 = p.add_instruction(migraphx::op::pooling{}, r2);
+    auto l6 = p.add_literal(migraphx::literal{b_shape, b});
+    auto c3 = p.add_instruction(migraphx::op::convolution{}, p1, l6);
+    auto l7 = p.add_literal(migraphx::literal{c_shape, c});
+    auto a3 = p.add_instruction(migraphx::op::add{}, c3, l7);
+    auto r3 = p.add_instruction(migraphx::op::relu{}, a3);
+    auto p3 = p.add_instruction(migraphx::op::pooling{}, r3);
+    p.add_instruction(migraphx::op::concat{1}, r1, p2, p3);
+    p.compile(horizontal_fusion_target{});
+    EXPECT(std::any_of(p.begin(), p.end(), [](const migraphx::instruction& ins) {
```

Comment:
In general I think we need to re-review some of our test cases that aren't 100% clear what exactly is going on. 

For this particular test case, I think the condition is rather weak of just checking whether or not a split exists in the program. When I run with MIGRAPHX_TRACE_COMPILE=1, I get the following result:
```
@0 = @literal{ ... } -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@1 = @literal{ ... } -> float_type, {64, 192, 1, 1}, {192, 1, 1, 1}
@2 = @literal{ ... } -> float_type, {1, 48, 35, 35}, {58800, 1225, 35, 1}
@3 = @literal{ ... } -> float_type, {48, 192, 1, 1}, {192, 1, 1, 1}
@4 = @literal{ ... } -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@5 = @literal{ ... } -> float_type, {64, 192, 1, 1}, {192, 1, 1, 1}
@6 = @literal{ ... } -> float_type, {1, 192, 35, 35}, {235200, 1225, 35, 1}
@7 = pooling[mode=average,padding=0, 0,padding_mode=0, 0,stride=1, 1,lengths=1, 1](@6) -> float_type, {1, 192, 35, 35}, {235200, 1225, 35, 1}
@8 = convolution[padding=0, 0,stride=1, 1,dilation=1, 1,padding_mode=0,group=1](@7,@5) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@9 = add(@8,@4) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@10 = relu(@9) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@11 = convolution[padding=0, 0,stride=1, 1,dilation=1, 1,padding_mode=0,group=1](@7,@3) -> float_type, {1, 48, 35, 35}, {58800, 1225, 35, 1}
@12 = add(@11,@2) -> float_type, {1, 48, 35, 35}, {58800, 1225, 35, 1}
@13 = relu(@12) -> float_type, {1, 48, 35, 35}, {58800, 1225, 35, 1}
@14 = pooling[mode=average,padding=0, 0,padding_mode=0, 0,stride=1, 1,lengths=1, 1](@13) -> float_type, {1, 48, 35, 35}, {58800, 1225, 35, 1}
@15 = convolution[padding=0, 0,stride=1, 1,dilation=1, 1,padding_mode=0,group=1](@7,@1) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@16 = add(@15,@0) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@17 = relu(@16) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@18 = pooling[mode=average,padding=0, 0,padding_mode=0, 0,stride=1, 1,lengths=1, 1](@17) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@19 = concat(@10,@14,@18) -> float_type, {1, 176, 35, 35}, {215600, 1225, 35, 1}


Pass: horizontal fusion
@0 = @literal{ ... } -> float_type, {1, 176, 35, 35}, {215600, 1225, 35, 1}
@1 = @literal{ ... } -> float_type, {176, 192, 1, 1}, {192, 1, 1, 1}
(6) @2 = @literal{ ... } -> float_type, {1, 192, 35, 35}, {235200, 1225, 35, 1}
(7) @3 = pooling[mode=average,padding=0, 0,padding_mode=0, 0,stride=1, 1,lengths=1, 1](@2) -> float_type, {1, 192, 35, 35}, {235200, 1225, 35, 1}
(8) @4 = convolution[padding=0, 0,stride=1, 1,dilation=1, 1,padding_mode=0,group=1](@3,@1) -> float_type, {1, 176, 35, 35}, {215600, 1225, 35, 1}
(9) @5 = add(@4,@0) -> float_type, {1, 176, 35, 35}, {215600, 1225, 35, 1}
(10) @6 = relu(@5) -> float_type, {1, 176, 35, 35}, {215600, 1225, 35, 1}
@7 = split(@6) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
@8 = split(@6) -> float_type, {137200}, {1}
@9 = load[offset=0,end=235200](@8) -> float_type, {1, 48, 35, 35}, {58800, 1225, 35, 1}
@10 = load[offset=235200,end=548800](@8) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
(14) @11 = pooling[mode=average,padding=0, 0,padding_mode=0, 0,stride=1, 1,lengths=1, 1](@9) -> float_type, {1, 48, 35, 35}, {58800, 1225, 35, 1}
(18) @12 = pooling[mode=average,padding=0, 0,padding_mode=0, 0,stride=1, 1,lengths=1, 1](@10) -> float_type, {1, 64, 35, 35}, {78400, 1225, 35, 1}
(19) @13 = concat(@7,@11,@12) -> float_type, {1, 176, 35, 35}, {215600, 1225, 35, 1}

[ COMPLETE ] test1
```
There are several transformations that occur, such as an overall reduction in instructions (fewer convolution, add, relu, literal) and changes to the input shapes. You may want to add additional checks and clarifications to this test case so that it is easier to revisit later on.

Reply:
fixed.